Update CURLOPT constant descriptions

haszi opened the pull request:
Updated descriptions of `CURLOPT_*` constants with the following:

- added some missing descriptions
- added type information on the associated accepted values
- added default values
- added cURL version information on availability, deprecation, removal 
- added applicable DocBook tags
- replaced lists of accepted value constants with their generalized version (eg.  `CURLALTSVC_*` instead of listing all `CURLALTSVC_` constants individually)
- standardized minor style details (e.g. use cURL instead of curl, libcurl, CURL; sentences to end in a period; remove double quotes from within `<literal>` tags, etc.)

Unfortunately since `CURLOPT` constants are the biggest subset of all cURL constants, the diff is pretty big.

cmb69 commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -68,8 +68,10 @@
     <para>
-     Pass the filename for cURL to use as the Alt-Svc cache file to read existing cache contents from and
-     possibly also write it back to a after a transfer, unless <constant>CURLALTSVC_READONLYFILE</constant>
+     Pass a <type>string</type> with the filename for cURL to use
+     as the Alt-Svc cache file to read existing cache contents from
+     and possibly also write it back to a after a transfer,
+     unless <constant>CURLALTSVC_READONLYFILE</constant>
```
It might be better not to mix line wrap changes with content changes (harder for translators).

haszi replied:
Should I rebase and remove all the line wrap changes or just do it in a follow-up commit on this PR?

cmb69 replied:
If the PR finally gets squashed merged, it doesn't matter; do whatever is easier for you. :)

cmb69 commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -127,10 +130,10 @@
    <listitem>
     <para>
-     Provides AWS V4 signature authentication on HTTP(S) header.
+     Provides AWS V4 signature authentication on HTTP(S) header as a <type>string</type>.
```
Could also use `&string;`, but `<type>string</type>` might be preferred nowadays.

cmb69 replied:
If I remember correctly, we deliberately decided to not use the entities in the function signatures (where `<type>string</type>` is supposed to occur often).

Maybe @Girgias knows whether it's preferable to use type elements or the respective entities elsewhere.

Girgias replied:
It is better to use the `<type>` element, as some translations translate the entity (don't ask me why doc-fr does it, but it does).

cmb69 replied:
With the same reasoning, they can translate the element as well. And soon we're back to
````
beschneide(zeichenkette $zeichenkette, zeichenkette $zeichen = " \n\r\t\v\x00"): zeichenkette
````

Girgias replied:
I'm pretty sure some of the QA scripts that can be used for translation check this. Usually the entity is better used when we want to talk about the "concept" of a string (or other type) rather then directly referring to it as a PHP type.

But the distinction is not useful most of the time. So maybe we should just fix translations that do that.

cmb69 commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -154,6 +158,8 @@
      The size of the buffer to use for each read. There is no guarantee
      this request will be fulfilled, however.
+     This option accepts any value that can be cast to a valid <type>int</type>.
+     Defaults to <constant>CURL_MAX_WRITE_SIZE</constant> (16kB).
```
Might be "dangerous" to include the actual size (16kB) here, since this might change upstream. At least maybe say
```suggestion
     Defaults to <constant>CURL_MAX_WRITE_SIZE</constant> (currently, 16kB).
```

cmb69 reviewed (commented):
Wow, a lot of work. Thank you!

Just left a couple of notes, before `CURLE_OPERATION_TIMEOUTED`. :)

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -378,7 +417,7 @@
    <listitem>
     <para>
      Pass a <type>string</type> naming a file with the concatenation of
-     CRL (Certificate Revocation List) (in PEM format)
+	 CRL (Certificate Revocation List) (in PEM format)
```
Seems to have an unintended indent change from spaces to tab here

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -614,6 +666,7 @@
     <para>
      Like <constant>CURLOPT_RANDOM_FILE</constant>, except a filename
      to an Entropy Gathering Daemon socket.
+     Available as of cURL 7.7 and deprecated as of cURL 7.84.0.
```
Maybe?
```suggestion
     Available as of cURL 7.7.0 and deprecated as of cURL 7.84.0.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -624,12 +677,17 @@
+      <member><literal>identity</literal></member>
+      <member><literal>deflate</literal></member>
+      <member><literal>gzip</literal></member>
+     </simplelist>
+     .
```
Doesn't this cause whitespace?

haszi replied:
Hmm. You're right, there is a space between the last item on the list and the period in the rendered file. I'll move the periods right after the `</simplelist>` tags.

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -654,8 +713,9 @@
-     is greater than or equal to 400. The default behavior is to return
+     is greater than or equal to <literal>400</literal>. The default behavior is to return
      the page normally, ignoring the code.
+     Available as of cURL 7.1.
```
Should this have a patch version?

haszi replied:
Good question. I've got all the version info from the [curl website](https://curl.se/libcurl/c/symbols-in-versions.html) and it seems that up until cURL 7.11.0, none of the major versions were using a patch version on the ".0" versions. Even the [release notes](https://curl.se/ch/7.10.html) refer to these versions with a patch version.

I'm OK with either leaving it as is or adding the ".0"s to these versions. Which one do you prefer?

Girgias replied:
I would prefer to add the `.0` to be consistent in our docs :)

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -795,6 +862,9 @@
      <constant>CURLOPT_TRANSFERTEXT</constant>. Use that instead.
+     Available as of cURL 7.1, deprecated as of cURL 7.11.1
+     and last available in cURL 7.15.5.
+     Removed as of PHP 7.3.
```
```suggestion
     Removed as of PHP 7.3.0.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -977,9 +1051,10 @@
-     to disable EPSV.
+     Set to &true; to first try an <literal>EPSV</literal> command for FTP
+     transfers before reverting back to <literal>PASV</literal>.
+     Set to &false; to disable <literal>EPSV</literal>.
+    Available as of cURL 7.9.2.
     </para>
```
Indent slightly off here

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1143,21 +1224,24 @@
+      <member><constant>CURLAUTH_AWS_SIGV4</constant></member>
+      <member><constant>CURLAUTH_ANY</constant></member>
+      <member><constant>CURLAUTH_ANYSAFE</constant></member>
+     </simplelist>
+     .
+     If more than one method is used, cURL will poll the server to see
```
Ditto whitespace concern?

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1380,16 +1471,20 @@
+      <member><literal>safe</literal></member>
+      <member><literal>confidential</literal></member>
+      <member><literal>private</literal></member>
+     </simplelist>
+     .
```
Ditto

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1695,9 +1818,11 @@
+     Use <constant>CURLOPT_RETURNTRANSFER</constant> instead.
+     Available as of cURL 7.1, deprecated as of cURL 7.8
+     and last available in cURL 7.15.5.
+     Removed as of PHP 7.3.
```
```suggestion
     Removed as of PHP 7.3.0.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -2025,8 +2168,8 @@
     <para>
-     A callback accepting five parameters.
-     The first is the cURL resource, the second is the total number of
+     A <type>callable</type> accepting five parameters and returning an <type>int</type>.
+     The first parameter is the cURL resource, the second is the total number of
      bytes expected to be downloaded in this transfer, the third is
      the number of bytes downloaded so far, the fourth is the total
```
Maybe this should use method synopsis?

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -2627,7 +2813,9 @@
-     A filename to be used to seed the random number generator for SSL.
+     A <type>string</type> with a filename to be used
+     to seed the random number generator for SSL.
+     Available as of cURL 7.7.0 and deprecated as of 7.84.0.
```
```suggestion
     Available as of cURL 7.7.0 and deprecated as of cURL 7.84.0.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -2665,14 +2855,15 @@
-     stream resource provided to cURL through the option
-     <constant>CURLOPT_INFILE</constant>, and the third is the maximum
-     amount of data to be read. The callback must return a string
+     A <type>callable</type> accepting three parameters and returning a <type>string</type>.
+     The first parameter is the cURL resource, the second is a
+     stream <type>resource</type> provided to cURL through the option
+     <constant>CURLOPT_READDATA</constant>, and the third is the maximum
+     amount of data to be read. The callback must return a <type>string</type>
```
Ditto methodsynopsis remark

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -3912,11 +4186,12 @@
-     The first is the cURL resource, and the second is a
-     string with the data to be written. The data must be saved by
-     this callback. It must return the exact number of bytes written
+     A <type>callable</type> accepting two parameters and returning an <type>int</type>.
+     The first parameter is the cURL resource, and the second is a
+     <type>string</type> with the data to be written. The data must be saved by
+     this callback. The callback must return the exact number of bytes written
      or the transfer will be aborted with an error.
```
Ditto method synopsys remark

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -3949,9 +4229,14 @@
-     A callback accepting two parameters.
-     Has a similar purpose as <constant>CURLOPT_PROGRESSFUNCTION</constant> but is more modern
-     and the preferred option from cURL.
+     A <type>callable</type> accepting five parameters and returning an <type>int</type>.
+     The first parameter is the cURL resource, the second is the total number of
+     bytes expected to be downloaded in this transfer, the third is
+     the number of bytes downloaded so far, the fourth is the total
+     number of bytes expected to be uploaded in this transfer, and the
+     fifth is the number of bytes uploaded so far.
+     Return <literal>1</literal> to abort the transfer
+     and set a <constant>CURLE_ABORTED_BY_CALLBACK</constant> error.
```
Ditto methodsynopsis remark

Girgias reviewed (changes requested):
There also seems to be missing a bunch of patch versions for cURL availability (mainly `7.1`)

haszi commented:
I think I have addressed all review comments such as:
- removed line wrap changes (all that I could find)
- added cURL .0 patch version numbers
- replaced callback descriptions with `<methodsynopsis>` elements 
- clarified that `CURLOPT_BUFFERSIZE` is *currently* 16KB (ie. can change)
- replaced tab with spaces
- added missing whitespace
- removed line break between `</simplelist>` tags and periods
- clarified that deprecation for `CURLOPT_RANDOM_FILE` is the cURL version number

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -198,8 +196,7 @@
     <para>
-     A <type>string</type> with a directory that holds multiple CA certificates.
-     Use this option alongside <constant>CURLOPT_SSL_VERIFYPEER</constant>.
+     A <type>string</type> with a directory that holds multiple CA certificates. Use this option alongside <constant>CURLOPT_SSL_VERIFYPEER</constant>.
```
Minor nit, line break at the full stop before "Use this option [...]".

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -256,8 +253,7 @@
     <para>
-     The number of milliseconds to wait while trying to connect.
-     Use <literal>0</literal> to wait indefinitely.
+     The number of milliseconds to wait while trying to connect. Use <literal>0</literal> to wait indefinitely.
```
Ditto line break

Girgias replied:
Ah was this a revert of a semantic line break?

Looking just from the previous review state was unclear ('thanks GitHub)

haszi replied:
Yeah, all these line break changes all reverts.

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -318,12 +313,10 @@
-     A <type>string</type> with the name of the file containing the cookie data.
-     The cookie file can be in Netscape format,
-     or just plain HTTP-style headers dumped into a file.
+     A <type>string</type> with the name of the file containing the cookie data. The cookie file can be in Netscape format, or just plain HTTP-style headers dumped into a file.
```
Line break

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1549,8 +1537,7 @@
      Can be used to set protocol specific login options, such as the
-     preferred authentication mechanism via <literal>AUTH=NTLM</literal>
-     or <literal>AUTH=*</literal>, and should be used in conjunction with the
+     preferred authentication mechanism via <literal>AUTH=NTLM</literal> or <literal>AUTH=*</literal>, and should be used in conjunction with the
```
Is the `*` in AUTH a literal or a placeholder?

haszi replied:
It is a literal and refers to any authorization method being acceptable.

Girgias reviewed (commented):
It looks like we got rid of some semantic breaks :/

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -191,8 +196,8 @@
     <para>
-     A directory that holds multiple CA certificates. Use this option
-     alongside <constant>CURLOPT_SSL_VERIFYPEER</constant>.
+     A <type>string</type> with a directory that holds multiple CA certificates. Use this option alongside <constant>CURLOPT_SSL_VERIFYPEER</constant>.
```
```suggestion
     A <type>string</type> with a directory that holds multiple CA certificates.
     Use this option alongside <constant>CURLOPT_SSL_VERIFYPEER</constant>.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -239,12 +253,13 @@
-     wait indefinitely.
-     <!-- http://curl.haxx.se/libcurl/c/curl_easy_setopt.html -->
-     If libcurl is built to use the standard system name resolver, that
+     The number of milliseconds to wait while trying to connect. Use <literal>0</literal> to wait indefinitely.
```
```suggestion
     The number of milliseconds to wait while trying to connect.
     Use <literal>0</literal> to wait indefinitely.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -297,11 +313,10 @@
-     be in Netscape format, or just plain HTTP-style headers dumped into
-     a file.
-     If the name is an empty string, no cookies are loaded, but cookie
+     A <type>string</type> with the name of the file containing the cookie data. The cookie file can be in Netscape format, or just plain HTTP-style headers dumped into a file.
```
```suggestion
     A <type>string</type> with the name of the file containing the cookie data.
     The cookie file can be in Netscape format, or just plain HTTP-style headers dumped into a file.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -312,8 +327,8 @@
     <para>
-     The name of a file to save all internal cookies to when the
-     handle's destructor is called.
+     A <type>string</type> with the name of a file to save all internal cookies to when the handle's destructor is called.
```
```suggestion
     A <type>string</type> with the name of a file to save all internal cookies to when
     the handle's destructor is called.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -459,8 +492,8 @@
     <para>
-     &true; to not allow URLs that include a username. Usernames are allowed by default (0).
-     Available as of PHP 7.3.0 and cURL 7.61.0
+     &true; to not allow URLs that include a username. Usernames are allowed by default.
+     Available as of PHP 7.3.0 and cURL 7.61.0.
```
```suggestion
     &true; to not allow URLs that include a username.
     Usernames are allowed by default.
     Available as of PHP 7.3.0 and cURL 7.61.0.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -496,9 +532,9 @@
-     Set the local IPv4 address that the resolver should bind to. The argument
-     should contain a single numerical IPv4 address as a string.
-     Available as of PHP 7.0.7 and cURL 7.33.0
+     Set the local IPv4 address that the resolver should bind to. The argument should contain a single numerical IPv4 address.
+     This option accepts a <type>string</type> or &null;.
```
```suggestion
     Set the local IPv4 address that the resolver should bind to.
     The argument should contain a single numerical IPv4 address.
     This option accepts a <type>string</type> or &null;.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -509,9 +545,9 @@
-     Set the local IPv6 address that the resolver should bind to. The argument
-     should contain a single numerical IPv6 address as a string.
-     Available as of PHP 7.0.7 and cURL 7.33.0
+     Set the local IPv6 address that the resolver should bind to. The argument should contain a single numerical IPv6 address.
+     This option accepts a <type>string</type> or &null;.
```
```suggestion
     Set the local IPv6 address that the resolver should bind to.
     The argument should contain a single numerical IPv6 address.
     This option accepts a <type>string</type> or &null;.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -666,8 +713,9 @@
-     The file that the transfer should be written to. The default
-     is <constant>STDOUT</constant> (the browser window).
+     Accepts a file handle <type>resource</type>
+     to the file that the transfer should be written to. The default is <constant>STDOUT</constant> (the browser window).
```
```suggestion
     to the file that the transfer should be written to.
     The default is <constant>STDOUT</constant> (the browser window).
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -817,12 +872,11 @@
-     string may be a plain IP address, a hostname, a network
-     interface name (under Unix), or just a plain '-' to use the
-     systems default IP address.
+     <type>string</type> may be a plain IP address, a hostname, a network interface name (under Unix), or just a plain <literal>-</literal> to use the system's default IP address.
```
```suggestion
     <type>string</type> may be a plain IP address, a hostname,
     a network interface name (under Unix),
     or just a plain <literal>-</literal> to use the system's default IP address.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -964,9 +1016,8 @@
-     &true; to use EPRT (and LPRT) when doing active
-     FTP downloads. Use &false; to disable EPRT and LPRT and use PORT
-     only.
+     Set to &true; to use <literal>EPRT</literal> (and <literal>LPRT</literal>) when doing active FTP downloads. Set to &false; to disable <literal>EPRT</literal> and <literal>LPRT</literal> and use <literal>PORT</literal> only.
```
```suggestion
     Set to &true; to use <literal>EPRT</literal> (and <literal>LPRT</literal>) when doing active FTP downloads.
     Set to &false; to disable <literal>EPRT</literal> and <literal>LPRT</literal> and use <literal>PORT</literal> only.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -977,9 +1028,8 @@
-     &true; to first try an EPSV command for FTP
-     transfers before reverting back to PASV. Set to &false;
-     to disable EPSV.
+     Set to &true; to first try an <literal>EPSV</literal> command for FTP transfers before reverting back to <literal>PASV</literal>. Set to &false; to disable <literal>EPSV</literal>.
```
```suggestion
     Set to &true; to first try an <literal>EPSV</literal> command for FTP transfers before reverting back to <literal>PASV</literal>.
     Set to &false; to disable <literal>EPSV</literal>.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1292,8 +1372,8 @@
     <para>
-     The name of the outgoing network interface to use. This can be an
-     interface name, an IP address or a host name.
+     Set to a <type>string</type> with the name of the outgoing network interface to use. This can be an interface name, an IP address or a host name.
```
```suggestion
     Set to a <type>string</type> with the name of the outgoing network interface to use.
     This can be an interface name, an IP address or a host name.
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1352,11 +1433,10 @@
     <para>
-     &true; to keep sending the request body if the HTTP code returned is
-     equal to or larger than 300. The default action would be to stop sending
+     Set to &true; to keep sending the request body if the HTTP code returned is equal to or larger than <literal>300</literal>. The default action would be to stop sending
```
```suggestion
     Set to &true; to keep sending the request body if the HTTP code returned is equal to or larger than <literal>300</literal>.
     The default action would be to stop sending
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -1695,9 +1798,12 @@
      the cURL functions.
-     Removed as of cURL 7.15.5; use <constant xmlns="http://docbook.org/ns/docbook">CURLOPT_RETURNTRANSFER</constant> instead.
+     Use <constant>CURLOPT_RETURNTRANSFER</constant> instead.
+     Available as of cURL 7.1, deprecated as of cURL 7.8
```
```suggestion
     Available as of cURL 7.1.0, deprecated as of cURL 7.8.0
```

Girgias commented on reference/curl/constants_curl_setopt.xml:
```diff
@@ -2713,8 +3025,8 @@
     <para>
-     The contents of the <literal>"Referer: "</literal> header to be used
-     in a HTTP request.
+     A <type>string</type> with the contents of the <literal>Referer: </literal> header to be used in a HTTP request.
```
```suggestion
     A <type>string</type> with the contents of the <literal>Referer: </literal>
     header to be used in a HTTP request.
```

Girgias reviewed (approved):
Readding semantic breaks where we are already editing it so it affects translations anyway
